compiler/rustc_resolve/src/late.rs

```diff
@@ -2688,28 +2726,42 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
         for item in trait_items {
             self.resolve_doc_links(&item.attrs, MaybeExported::Ok(item.id));
             match &item.kind {
-                AssocItemKind::Const(box ast::ConstItem { ty, expr, .. }) => {
-                    self.visit_ty(ty);
-                    // Only impose the restrictions of `ConstRibKind` for an
-                    // actual constant expression in a provided default.
-                    if let Some(expr) = expr {
-                        // We allow arbitrary const expressions inside of associated consts,
-                        // even if they are potentially not const evaluatable.
-                        //
-                        // Type parameters can already be used and as associated consts are
-                        // not used as part of the type system, this is far less surprising.
-                        self.with_lifetime_rib(
-                            LifetimeRibKind::Elided(LifetimeRes::Infer),
-                            |this| {
-                                this.with_constant_rib(
-                                    IsRepeatExpr::No,
-                                    ConstantHasGenerics::Yes,
-                                    None,
-                                    |this| this.visit_expr(expr),
-                                )
-                            },
-                        );
-                    }
+                // FIXME(generic_consts): Reuse `walk_assoc_item` here somehow?
```

Comment:
Is `with_constant_rib` really useful with `IsRepeatExpr::No, ConstantHasGenerics::Yes`? I'm not even sure it has any effect. Perhaps just upvar diagnostics. If it doesn't, `walk_assoc_item` can be reused.

Reply:
I've checked `tests/ui` after having blatantly ripped out any `with_constant_rib(No, Yes)` and it does indeed regress the quality of diagnostics, quite heavily actually:

```diff
- error[E0435]: attempt to use a non-constant value in a constant
+ error[E0434]: can't capture dynamic environment in a fn item
```

Reply:
I'll try a refactor on my own, once your PR lands.